optimum/onnx/graph_transformations.py

```diff
@@ -79,3 +96,219 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
+                _infer_output_shape(model_output_2),
+            )
+            model1.graph.output.insert(idx, new_output)
+
+    if not all(
+        model_output_1 == model_output_2
+        for model_output_1, model_output_2 in zip(model1.graph.output, model2.graph.output)
+    ):
+        raise RuntimeError(f"Failed to unify outputs of given ONNX model protos.")
+
+
+def _get_all_inputs(model_list: List[ModelProto]):
+    inputs = []
+    input_names = []
+    for model in model_list:
+        for input in model.graph.input:
+            if input.name not in input_names:
+                input_names.append(input.name)
+                inputs.append(input)
+    return inputs
+
+
+def _get_onnx_opset(model: ModelProto):
+    opset_import = model.opset_import[0]
+    return getattr(opset_import, "version")
+
+
+def simplify_onnx_graph(model: ModelProto, print_info: bool = True) -> ModelProto:
+
+    model_simp, check = simplify(model)
+    assert check, "Simplified ONNX model could not be validated."
+    if print_info:
+        print_simplifying_info(model, model_simp)
+
+    return model_simp
+
+
+def _deduplicated_cross_model_initializers(models: List[ModelProto], suffix: str = None):
+
+    all_initializers = list()
+    for model in models:
+        all_initializers += list(model.graph.initializer)
+
+    duplicates = _find_duplicate_initializers(all_initializers)
+    name_sharing_dict = _create_name_sharing_dict(duplicates, suffix=suffix)
+    for model in models:
+        _replace_input_names(model, name_sharing_dict)
+
+    deduplicated_initializers = list()
+    deduplicated_name = list()
+
+    for initializer in all_initializers:
+        if name_sharing_dict[initializer.name] not in deduplicated_name:
+            deduplicated_name.append(name_sharing_dict[initializer.name])
+            initializer.name = name_sharing_dict[initializer.name]
+            deduplicated_initializers.append(initializer)
+
+    return deduplicated_initializers
+
+
+def merge_decoders(
+    decoder: ModelProto,
+    decoder_with_past: ModelProto,
+    graph_name: str = "merged",
+    producer_name: str = "optimum-onnx",
+    simplify_graph: bool = False,
+) -> ModelProto:
+    """
+    Fuses decoder ONNX model and decoder with past ONNX model into one ONNX model with if logic.
+
+    Args:
+        decoder (`~onnx.ModelProto`): Decoder ONNX model.
+        decoder_with_past (`~onnx.ModelProto`): Decoder with past ONNX model.
+        graph_name (`~str`): Name of the parent graph(graph of the control flow node).
+        producer_name (`~str`): Graph producer name.
+        simplify_graph (`~bool`): Whether to use onnx-simplifier to simplify subgraphs.
+
+    Returns:
+        `~onnx.ModelProto`: The fused decoder ONNX model.
+    """
+    if simplify_graph:
+        decoder = simplify_onnx_graph(decoder)
+        decoder_with_past = simplify_onnx_graph(decoder_with_past)
+
+    _unify_onnx_outputs(decoder, decoder_with_past)
```

Comment:
just to understand, why do we do this? Is it so that the decoder spits out the past key values? I think a comment would help

Reply:
According to the rules of "If" op, two subgraphs must have the same number of outputs. Comment added to this function.